sdk/core/core-rest-pipeline/src/util/tokenCycler.ts

```diff
@@ -52,50 +44,61 @@ export const DEFAULT_CYCLER_OPTIONS: TokenCyclerOptions = {
 };
 
 /**
- * Converts an an unreliable access token getter (which may resolve with null)
- * into an AccessTokenGetter by retrying the unreliable getter in a regular
- * interval.
+ * Calls to asyncFn until the timeout is reached.
+ * Prevents errors from bubbling up until the refresh timeout is reached.
+ * Once the timeout is reached, errors will bubble up, and null return values will throw with the received "errorMessage".
  *
- * @param getAccessToken - A function that produces a promise of an access token that may fail by returning null.
- * @param retryIntervalInMs - The time (in milliseconds) to wait between retry attempts.
- * @param refreshTimeout - The timestamp after which the refresh attempt will fail, throwing an exception.
- * @returns - A promise that, if it resolves, will resolve with an access token.
+ * @param asyncFn - The asynchronous function to call.
+ * @param timeout - Unix time when to stop ignoring errors or null values.
+ * @param errorMessage - Message to be used if the final value after the timeout is reached is still "null".
+ * @returns - A promise that can be resolved with T or null until the time is reached, after which it will throw if it can't resolve with a value of the type T.
  */
-async function beginRefresh(
-  getAccessToken: () => Promise<AccessToken | null>,
-  retryIntervalInMs: number,
-  refreshTimeout: number
-): Promise<AccessToken> {
-  // This wrapper handles exceptions gracefully as long as we haven't exceeded
-  // the timeout.
-  async function tryGetAccessToken(): Promise<AccessToken | null> {
-    if (Date.now() < refreshTimeout) {
-      try {
-        return await getAccessToken();
-      } catch {
-        return null;
-      }
-    } else {
-      const finalToken = await getAccessToken();
-
-      // Timeout is up, so throw if it's still null
-      if (finalToken === null) {
-        throw new Error("Failed to refresh access token.");
-      }
-
-      return finalToken;
+async function tryUntilTimeout<T>(
```

Comment:
To me the names `tryUntilTimeout` and `retryUntilTimeout` mean the same thing. This function is specifically made to be part of the inner loop of `retryUntilTimeout`, and I think it has limited utility on its own.

Reply:
Inside of retryUntilTimeout I think it adds too much complexity for one function. The name could be better though. This could become something like "run silently unless a condition is met", and receive a condition instead of a timeout, which would be generic.

Reply:
I just think it's a very particular function. "tryOnceAndThrowIfTimeoutExpired"... It only makes sense to be used with `retryUntilTimeout`, which still parses in my brain as meaning the exact same thing as `tryUntilTimeout`. Maybe this function could simply be called `tryOnce`.

I generally like the idea of accepting a predicate representing the error condition instead of checking the date, but like I mentioned in the other comment, I'm skeptical that this makes much sense to use with other "refreshable things" as long as `| null` is baked so deeply in. That feels like a leaky abstraction where things would start wanting to return null as a failure mode just for the sake of fitting into that abstraction, even if it's more natural to reject/call a handler, etc.

It might also be possible to use some clever promise chains to eliminate this secondary function entirely.

Reply:
Oh I see! I like that. Alright, let's come back to this later when we have more time 👍